crates/ra_ide_api/src/syntax_highlighting.rs

```diff
@@ -32,59 +33,74 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
 
                     // FIXME: try to reuse the SourceAnalyzers
                     let analyzer = hir::SourceAnalyzer::new(db, file_id, name_ref.syntax(), None);
                     match classify_name_ref(db, &analyzer, name_ref) {
-                        Some(Method(_)) => "function",
-                        Some(Macro(_)) => "macro",
-                        Some(FieldAccess(_)) => "field",
-                        Some(AssocItem(ImplItem::Method(_))) => "function",
-                        Some(AssocItem(ImplItem::Const(_))) => "constant",
-                        Some(AssocItem(ImplItem::TypeAlias(_))) => "type",
-                        Some(Def(ModuleDef::Module(_))) => "module",
-                        Some(Def(ModuleDef::Function(_))) => "function",
-                        Some(Def(ModuleDef::Struct(_))) => "type",
-                        Some(Def(ModuleDef::Union(_))) => "type",
-                        Some(Def(ModuleDef::Enum(_))) => "type",
-                        Some(Def(ModuleDef::EnumVariant(_))) => "constant",
-                        Some(Def(ModuleDef::Const(_))) => "constant",
-                        Some(Def(ModuleDef::Static(_))) => "constant",
-                        Some(Def(ModuleDef::Trait(_))) => "type",
-                        Some(Def(ModuleDef::TypeAlias(_))) => "type",
-                        Some(SelfType(_)) => "type",
-                        Some(Pat(_)) => "text",
-                        Some(SelfParam(_)) => "type",
-                        Some(GenericParam(_)) => "type",
-                        None => "text",
+                        Some(Method(_)) => ("function", None),
+                        Some(Macro(_)) => ("macro", None),
+                        Some(FieldAccess(_)) => ("field", None),
+                        Some(AssocItem(ImplItem::Method(_))) => ("function", None),
+                        Some(AssocItem(ImplItem::Const(_))) => ("constant", None),
+                        Some(AssocItem(ImplItem::TypeAlias(_))) => ("type", None),
+                        Some(Def(ModuleDef::Module(_))) => ("module", None),
+                        Some(Def(ModuleDef::Function(_))) => ("function", None),
+                        Some(Def(ModuleDef::Struct(_))) => ("type", None),
+                        Some(Def(ModuleDef::Union(_))) => ("type", None),
+                        Some(Def(ModuleDef::Enum(_))) => ("type", None),
+                        Some(Def(ModuleDef::EnumVariant(_))) => ("constant", None),
+                        Some(Def(ModuleDef::Const(_))) => ("constant", None),
+                        Some(Def(ModuleDef::Static(_))) => ("constant", None),
+                        Some(Def(ModuleDef::Trait(_))) => ("type", None),
+                        Some(Def(ModuleDef::TypeAlias(_))) => ("type", None),
+                        Some(SelfType(_)) => ("type", None),
+                        Some(Pat(ptr)) => ("variable", Some(hash(ptr.syntax_node_ptr().range()))),
```

Comment:
Hashing position is not ideal: the colors would change after every modification.

I think we should hash the name of variable. Note that this will cause all shadowed versions of var to be colored the same, but I think that's ok as a protopyte. To deal with shadowing, we can hash a tuple of `(name, N)`, where N means that this Nth definition of name in scope (so, `(foo, 0)` for first one, `(foo, 1)` for first shadowed, etc)